File: javascript/util/phi.js
const primal = require(`./primal`);

function phi(val, primes) {
  if (primal.isPrime(val, primes)) {
    return val - 1;
  }

  const primeFactors = primal.getPrimeFactors(val, primes);
  const count = val * primeFactors.reduce((a, c) => a * (1 - 1 / c.base), 1);
  return count;
}

module.exports = phi;

File: python/util/phi.py
from . import primal


def get_phi(val, primes):
  if primal.is_prime(val, primes):
    return val - 1

  prime_factors = primal.get_prime_factors(val, primes)
  multiplier = 1
  for prime_factor in prime_factors:
    multiplier *= 1 - 1 / prime_factor['base']

  count = val * multiplier

  return count